### ISSUE TITLE
[python - Rips] `create_from_array` should be promoted for Rips construction

### ISSUE THREAD

**Issue by VincentRouvreau** [GUDHI/gudhi-devel#1169]
When constructing a Rips from a distance matrix, there is a big benchmark difference between RipsComplex interface and create_from_array:
```python
from timeit import default_timer as timer
from datetime import timedelta
import gudhi
from gudhi.datasets.generators import points

points = points.sphere(n_samples = 100, ambient_dim = 3, radius = 1, sample = "random")
distance_matrix = cdist(points, points)
start = timer(); gudhi.RipsComplex(distance_matrix=distance_matrix).create_simplex_tree(); print(timedelta(seconds=timer()-start))
# 0:00:00.003350
start = timer(); gudhi.SimplexTree.create_from_array(distance_matrix); print(timedelta(seconds=timer()-start))
# 0:00:00.000359
```

Thanks @DavidLapous and @jumdc for the report

**Comment by mglisse** [comment]
Let's split it in 2 parts
```python
%time r = gudhi.RipsComplex(distance_matrix=distance_matrix)
%time s1 = r.create_simplex_tree()
%time s2 = gudhi.SimplexTree.create_from_array(distance_matrix)
```
> CPU times: user 962 μs, sys: 31 μs, total: 993 μs
Wall time: 995 μs
CPU times: user 268 μs, sys: 0 ns, total: 268 μs
Wall time: 206 μs
CPU times: user 89 μs, sys: 0 ns, total: 89 μs
Wall time: 89.4 μs

So the slowest part is init_matrix.

* The 2 functions don't exactly take the same argument, RipsComplex accepts any iterable (though it is possible to detect if the input is a 2d array and do something different in that case)
* RipsComplex first converts to vector of vectors, then to boost graph, and then to simplex tree, while I think create_from_array has no intermediate representation (it might copy the matrix first if it is uses a type other than float64).

Out of curiosity, does the construction of the Rips complex dominate your running time, or did you just happen to notice this but it doesn't have a real impact for you because there are much more expensive operations around it?

**Comment by DavidLapous** [comment]
> Ah, H0... We shouldn't even need a Simplex_tree for that, the array is a nice representation of a graph. I don't know about multiparameter, but for single parameter, it should not be hard to compute H0 without a Simplex_tree (I think RipsPersistence already does it in master, though it might still copy the array to its own format, I don't remember). Would it make sense for your use case to try and bypass the simplex tree?

Yes; good thing to know it's in master

> Don't you have to call .cpu() and .detach() before you can pass it to Gudhi? 

not if you do a 
```python
with torch.no_grad():
  # stuff
```
before. 

> I think that would probably give bad performance, although I don't know what it would look like exactly.

Not really, $O(n\log n)$ for a sort (which is already needed for persistence) + $O(n\log n)$ to get the coordinates + $O(n)$ to reevaluate the barcode. Furthermore, everything can be done in "Array api" syntax (i.e., efficiently and in a parallel fashion in python).

> I don't remember if anyone has tried computing H0 on a GPU, most efforts I know about are about H1+. It wouldn't need to be faster than the CPU, just avoid having to go back to the CPU (plus it might help with autodiff).

For H0 there are definitely some work on the minimum spanning tree which should yield a significative improvement, e.g., https://dl.acm.org/doi/10.1145/3581784.3607093, with the code here https://github.com/alexfallin/MST-Artifact-SC23/.
